### PR TITLE
Make function clauselist_selectivity have same format with upstream postgres

### PR DESCRIPTION
**Pull request by jingwen-yang-yjw**
This pr includes two commits.

**Commit 1:**
Function `clauselist_selectivity` in gpdb7 has one more parameter `use_damping` than upstream postgres.
It causes many postgres's fdw compilation failures in gpdb7(like kafka_fdw, clickhouse_fdw and arrow_fdw).

Because function `clauselist_selectivity` is a general function called by most postgres's fdw,
it's better to keep function `clauselist_selectivity` having same format with upstream postgres.

In this commit, we rename function `clauselist_selectivity` to `clauselist_selectivity_extended`  for gpdb7 internal use.
Meanwhile, we keep function `clauselist_selectivity` having same format with upstream postgres for external fdw use.

**Commit 2:**
We add function `tag_hash` declaration in src/include/utils/hsearch.h.
For function `tag_hash` declaration, gpdb7 is not consistent with postgres 12, but with a higher version.
It causes mysql_fdw compilation failures in gpdb7.

To solve this issue, we add function `tag_hash` declaration in src/include/utils/hsearch.h.

### CODE REVIEW

File: src/backend/optimizer/path/clausesel.c [hunk 1]
```diff
@@ -75,13 +75,42 @@ cmpSelectivity
...
+ *	  it allows caching of results. 
+ *
+ * clauselist_selectivity is kept same as upstream here. 
+ * Because this function is called by most fdw (Foreign Data Wrapper) extensions.
+ * This functions just simply call clauselist_selectivity_use_damping() and set use_damping to false.
+ * 
+ * Greenplum specific behavior:
+ * Greenplum calls clauselist_selectivity_use_damping() directly.
+ * Function clauselist_selectivity() is kept for external fdw.
+ *
```
Comment (zhaorui20005) [inline src/backend/optimizer/path/clausesel.c:93]: Please keep the indent tidy.

File: src/backend/optimizer/path/clausesel.c [hunk 2]
```diff
@@ -75,13 +75,42 @@ cmpSelectivity
 /****************************************************************************
  *		ROUTINES TO COMPUTE SELECTIVITIES
  ****************************************************************************/
-
```
[the remark this answers is not in the document]
Comment (jingwen-yang-yjw) [reply src/backend/optimizer/path/clausesel.c:78]: Thank you.
I'll fix it.

File: src/backend/optimizer/path/clausesel.c (hunk 1 above)
Comment (jingwen-yang-yjw) [reply src/backend/optimizer/path/clausesel.c:93]: Thank you.
I'll check and keep the indent tidy.

File: src/include/optimizer/optimizer.h [hunk 3]
```diff
@@ -61,6 +61,11 @@ extern Selectivity clause_selectivity(PlannerInfo *root,
 									  SpecialJoinInfo *sjinfo,
 									  bool use_damping);
 extern Selectivity clauselist_selectivity(PlannerInfo *root,
+										  List *clauses,
+										  int varRelid,
+										  JoinType jointype,
+										  SpecialJoinInfo *sjinfo);									  
+extern Selectivity clauselist_selectivity_use_damping(PlannerInfo *root,
```
Comment (huiliang-liu) [inline src/include/optimizer/optimizer.h:68]: `clauselist_selectivity_use_damping` usually means the `use_damping` is true. I think this name may make people feel confused. could we rename it as `clauselist_selectivity_standard` ? to compare with `clauselist_selectivity_simple` in line 74.

Comment (jingwen-yang-yjw) [reply src/include/optimizer/optimizer.h:68]: Yeah. I think you're right.
I will rename clauselist_selectivity_use_damping.

**Review by zhaorui20005 — APPROVED** [review]

**Review by huiliang-liu — APPROVED** [review]

**Pull request merged**